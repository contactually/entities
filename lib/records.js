'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.recordsFromFieldDefinitions = exports.entitiesToRecords = exports.dependentEntityKeys = undefined;

var _each2 = require('lodash/each');

var _each3 = _interopRequireDefault(_each2);

var _upperFirst2 = require('lodash/upperFirst');

var _upperFirst3 = _interopRequireDefault(_upperFirst2);

var _reduce2 = require('lodash/reduce');

var _reduce3 = _interopRequireDefault(_reduce2);

var _immutable = require('immutable');

var _IterableSchema = require('normalizr/lib/IterableSchema');

var _IterableSchema2 = _interopRequireDefault(_IterableSchema);

var _EntitySchema = require('normalizr/lib/EntitySchema');

var _EntitySchema2 = _interopRequireDefault(_EntitySchema);

var _index = require('./index');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var recordizeProperties = function recordizeProperties(memo, defaultValue, property) {
  if (defaultValue && defaultValue.constructor === Object) {
    var NestedRecord = new _immutable.Record((0, _reduce3.default)(defaultValue, recordizeProperties, {}));
    memo[property] = NestedRecord().mergeDeep(NestedRecord(defaultValue));
  } else {
    memo[property] = defaultValue;
  }

  return memo;
};

var recordsFromFieldDefinitions = function recordsFromFieldDefinitions(fieldDefinitions) {
  return (0, _reduce3.default)(fieldDefinitions, function (memo, properties, type) {
    var className = (0, _upperFirst3.default)(type);
    var recordizedProperties = (0, _reduce3.default)(properties, recordizeProperties, {});

    memo[className] = new _immutable.Record(recordizedProperties, className);

    return memo;
  }, {});
};

/**
 * @description
 * Iterates through the given normalizedEntityMap recursively, making sure that all
 * subkeys are Records.
 * @example
 * entitiesToRecords({
 *   contact: {
 *     contact_123: {id: 'contact_123', ...}
 *   },
 *   bucket: {
 *     bucket_123: {id: 'bucket_123', ...}
 *   }
 * })
 * => {
 *   contact: {
 *     contact_123: records.Contact(...)
 *   },
 *   bucket: {
 *     bucket_123: records.Bucket(...)
 *   }
 * }
 * @param {NormalizedEntityMap} normalizedEntityMap
 * @returns {NormalizedEntityMap}
 */
var entitiesToRecords = function entitiesToRecords(normalizedEntityMap) {
  return (0, _immutable.fromJS)(normalizedEntityMap).withMutations(function (entities) {
    return entities.forEach(function (entitiesOfType, type) {
      var RecordClass = _index.EntitiesConfig.records[(0, _upperFirst3.default)(type)];

      if (!RecordClass) return;

      entitiesOfType.forEach(function (entity, id) {
        if (!(entity instanceof RecordClass)) {
          entities.setIn([type, id], RecordClass().mergeDeep(RecordClass(entity)));
        }
      });
    });
  });
};

/**
 * @description
 * Recurses through a schema, reducing a list of all involved/nested entity types.
 * @example
 * const contact = new Schema('contact')
 * const bucket = new Schema('bucket')
 * const bucketPermission = new Schema('bucketPermission')
 *
 * contact.define({ contacts: arrayOf(contact) })
 * bucket.define({ bucketPermissions: arrayOf(bucketPermission) })
 *
 * dependentEntityKeys(contact)
 * => ['contact', 'bucket', 'bucketPermission']
 * dependentEntityKeys(bucket)
 * => ['bucket', 'bucketPermission']
 * dependentEntityKeys(bucketPermission)
 * => ['bucketPermission']
 * @param {EntitySchema} schema
 * @param {string[]=} keys
 * @returns {string[]}
 */
var dependentEntityKeys = function dependentEntityKeys(schema) {
  var keys = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];

  var schemaKey = schema.getKey();

  if (keys.includes(schemaKey)) {
    return keys;
  }

  keys.push(schemaKey);

  (0, _each3.default)(schema, function (value, key) {
    if (value instanceof _EntitySchema2.default) {
      dependentEntityKeys(value, keys);
    } else if (value instanceof _IterableSchema2.default) {
      dependentEntityKeys(value.getItemSchema(), keys);
    }
  });

  return keys;
};

exports.dependentEntityKeys = dependentEntityKeys;
exports.entitiesToRecords = entitiesToRecords;
exports.recordsFromFieldDefinitions = recordsFromFieldDefinitions;