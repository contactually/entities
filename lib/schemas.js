'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schemasFromFieldDefinitions = undefined;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _normalizr = require('normalizr');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Creates a Schema for each of the given record types.
 */
function schemasFromFieldDefinitions(fieldDefinitions, schemaDefinitions) {
  var allSchemaTypes = (0, _keys2.default)(fieldDefinitions).concat((0, _keys2.default)(schemaDefinitions));
  return allSchemaTypes.reduce(function (memo, schemaType) {
    memo[schemaType] = new _normalizr.Schema(schemaType, schemaDefinitions[schemaType]);
    return memo;
  }, {});
}

exports.schemasFromFieldDefinitions = schemasFromFieldDefinitions;