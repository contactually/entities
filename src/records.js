/* @flow */
import * as _ from 'lodash'
import { Record, fromJS, Map } from 'immutable'

import IterableSchema from 'normalizr/lib/IterableSchema'
import EntitySchema from 'normalizr/lib/EntitySchema'

import type {NormalizedEntityMap} from './module/selectors'

import EntitiesConfig from './Config'

type Records = {
  [entityName: string]: Record<{}>
}

const recordizeProperties = (memo, defaultValue, property) => {
  if (defaultValue && defaultValue.constructor === Object) {
    const NestedRecord = Record(_.reduce(defaultValue, recordizeProperties, {}))
    memo[property] = NestedRecord().mergeDeep(NestedRecord(defaultValue))
  } else {
    memo[property] = defaultValue
  }

  return memo
}

const recordsFromFieldDefinitions = (fieldDefinitions: any) => {
  return _.reduce(fieldDefinitions, (memo, properties, type) => {
    const className = _.upperFirst(type)
    const recordizedProperties = _.reduce(properties, recordizeProperties, {})

    memo[className] = Record(recordizedProperties, className)

    return memo
  }, {})
}

/**
 * @description
 * Iterates through the given normalizedEntityMap recursively, making sure that all
 * subkeys are Records.
 * @example
 * entitiesToRecords({
 *   contact: {
 *     contact_123: {id: 'contact_123', ...}
 *   },
 *   bucket: {
 *     bucket_123: {id: 'bucket_123', ...}
 *   }
 * })
 * => {
 *   contact: {
 *     contact_123: records.Contact(...)
 *   },
 *   bucket: {
 *     bucket_123: records.Bucket(...)
 *   }
 * }
 * @param {NormalizedEntityMap} normalizedEntityMap
 * @returns {NormalizedEntityMap}
 */
const entitiesToRecords = (normalizedEntityMap: NormalizedEntityMap): NormalizedEntityMap => (
  fromJS(normalizedEntityMap).withMutations(
    (entities: NormalizedEntityMap) => entities.forEach((entitiesOfType: Map<string, Record<{}>>, type: string) => {
      const RecordClass = EntitiesConfig.records[_.upperFirst(type)]

      if (!RecordClass) return

      entitiesOfType.forEach((entity, id) => {
        if (!(entity instanceof RecordClass)) {
          // $FlowFixMe
          entities.setIn([type, id], RecordClass().mergeDeep(RecordClass(entity)))
        }
      })
    })
  )
)

/**
 * @description
 * Recurses through a schema, reducing a list of all involved/nested entity types.
 * @example
 * const contact = new Schema('contact')
 * const bucket = new Schema('bucket')
 * const bucketPermission = new Schema('bucketPermission')
 *
 * contact.define({ contacts: arrayOf(contact) })
 * bucket.define({ bucketPermissions: arrayOf(bucketPermission) })
 *
 * dependentEntityKeys(contact)
 * => ['contact', 'bucket', 'bucketPermission']
 * dependentEntityKeys(bucket)
 * => ['bucket', 'bucketPermission']
 * dependentEntityKeys(bucketPermission)
 * => ['bucketPermission']
 * @param {EntitySchema} schema
 * @param {string[]=} keys
 * @returns {string[]}
 */
const dependentEntityKeys = (schema: EntitySchema, keys: string[] = []): string[] => {
  const schemaKey = schema.getKey()

  if (keys.includes(schemaKey)) {
    return keys
  }

  keys.push(schemaKey)

  _.each(schema, (value, key) => {
    if (value instanceof EntitySchema) {
      dependentEntityKeys(value, keys)
    } else if (value instanceof IterableSchema) {
      dependentEntityKeys(value.getItemSchema(), keys)
    }
  })

  return keys
}

export {
  dependentEntityKeys,
  entitiesToRecords,
  recordsFromFieldDefinitions
}
