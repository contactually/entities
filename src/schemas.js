import * as normalizr from 'normalizr'

type Schemas = {
  [entityName: string]: Schema
}

type FieldDefinitions = {
  [entityName: string]: {
    [propertyName: string]: any
  }
}

/**
 * Creates a Schema for each of the given record types.
 */
function schemasFromFieldDefinitions (fieldDefinitions: FieldDefinitions): Schemas {
  const allSchemaTypes = Object.keys(fieldDefinitions)
  return allSchemaTypes.reduce((memo: Schemas, schemaType: string) => {
    memo[schemaType] = new normalizr.schema.Entity(schemaType)
    return memo
  }, {})
}

export {
  schemasFromFieldDefinitions
}
