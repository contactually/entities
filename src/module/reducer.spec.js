import { schema } from 'normalizr'
import { Record, OrderedMap, Map, fromJS } from 'immutable'
import EntitiesConfig from '../Config'
import {recordsFromFieldDefinitions} from '../records'
import {schemasFromFieldDefinitions} from '../schemas'
import {
  handleAddActiveRequest,
  handleDeleteActiveRequest,
  handleMergeContainerData,
  handleMergeFilters,
  handleSetFilters,
  handleResetContainerData,
  handleMergeEntities,
  handleSetRequestStarted,
  handleSetRequestCompleted
} from './reducer'

describe('reducer', () => {
  describe('handleMergeEntities', () => {
    test('preserves order of nested maps', () => {
      const fieldDefinitions = {
        importedCsv: {
          importOptions: {
            columnMap: OrderedMap()
          }
        }
      }

      const records = recordsFromFieldDefinitions(fieldDefinitions)
      const schemas = schemasFromFieldDefinitions(fieldDefinitions)

      EntitiesConfig.configure({
        records,
        schemas
      })

      const state = Map({
        entities: Map({
          contact: Map({})
        })
      })
      const entities = {
        importedCsv: {
          imported_csv_123: {
            importOptions: {
              columnMap: {
                'fName': {
                  'skip': false,
                  'mappedHeader': 'custom_field',
                  'exampleData': [
                    'Olive',
                    'austin',
                    'who'
                  ]
                },
                'lName': {
                  'skip': false,
                  'mappedHeader': 'custom_field',
                  'exampleData': [
                    'What',
                    'When'
                  ]
                },
                'eMail': {
                  'skip': false,
                  'mappedHeader': 'custom_field',
                  'exampleData': [
                    'olivecarroll@gmail.com',
                    'bad email dawg'
                  ]
                },
                'email2': {
                  'skip': false,
                  'mappedHeader': 'email',
                  'exampleData': [
                    'olivecarroll@yahoo.com'
                  ]
                },
                'phoneNumber': {
                  'skip': false,
                  'mappedHeader': 'phone',
                  'exampleData': [
                    '(123) 456-7890',
                    '123 333 4444'
                  ]
                },
                'title': {
                  'skip': false,
                  'mappedHeader': 'title',
                  'exampleData': [
                    'CEO',
                    'idk'
                  ]
                },
                'company': {
                  'skip': false,
                  'mappedHeader': 'company',
                  'exampleData': [
                    'Canines Inc'
                  ]
                },
                'website': {
                  'skip': false,
                  'mappedHeader': 'website',
                  'exampleData': [
                    'www.olive.com'
                  ]
                },
                'homeStreet': {
                  'skip': false,
                  'mappedHeader': 'address_line_1',
                  'type': 'Home',
                  'exampleData': [
                    '123 Main Street'
                  ]
                },
                'homeCity': {
                  'skip': false,
                  'mappedHeader': 'address_city',
                  'type': 'Home',
                  'exampleData': [
                    'Arlington'
                  ]
                },
                'homeState': {
                  'skip': false,
                  'mappedHeader': 'address_state',
                  'type': 'Home',
                  'exampleData': [
                    'VA'
                  ]
                },
                'homeZip': {
                  'skip': false,
                  'mappedHeader': 'address_zip',
                  'type': 'Home',
                  'exampleData': [
                    '20001'
                  ]
                },
                'bucket': {
                  'skip': false,
                  'mappedHeader': 'buckets',
                  'exampleData': [
                    'Past Client'
                  ]
                },
                'tag': {
                  'skip': false,
                  'mappedHeader': 'tags',
                  'exampleData': [
                    'Labrador'
                  ]
                }
              }
            }
          }
        }
      }
      const result = handleMergeEntities(state, { entities })
      const keys = result
        .getIn(['entities', 'importedCsv', 'imported_csv_123', 'importOptions', 'columnMap']).keySeq().toArray()

      expect(keys).toMatchObject([
        'fName',
        'lName',
        'eMail',
        'email2',
        'phoneNumber',
        'title',
        'company',
        'website',
        'homeStreet',
        'homeCity',
        'homeState',
        'homeZip',
        'bucket',
        'tag'
      ])
    })
  })
})
