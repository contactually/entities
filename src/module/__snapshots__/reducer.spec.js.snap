// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer handleMergeEntities preserves order of nested maps 1`] = `
Array [
  "a_key",
  "z_key",
  "h_key",
  "b_key",
  "c_key",
  "d_key",
  "e_key",
  "aa_key",
]
`;
